File: Sources/Sentry/SentryCrashInstallationReporter.m
#import "SentryCrashInstallationReporter.h"
#import "SentryCrash.h"
#import "SentryCrashInstallation+Private.h"
#import "SentryCrashReportSink.h"
#import "SentryDefines.h"
#import "SentryLog.h"

NS_ASSUME_NONNULL_BEGIN

@interface
SentryCrashInstallationReporter ()

@property (nonatomic, strong) SentryInAppLogic *inAppLogic;
@property (nonatomic, strong) SentryCrashWrapper *crashWrapper;

@end

@implementation SentryCrashInstallationReporter

- (instancetype)initWithInAppLogic:(SentryInAppLogic *)inAppLogic
                      crashWrapper:(SentryCrashWrapper *)crashWrapper
{
    if (self = [super initWithRequiredProperties:[NSArray new]]) {
        self.inAppLogic = inAppLogic;
        self.crashWrapper = crashWrapper;
    }
    return self;
}

- (id<SentryCrashReportFilter>)sink
{
    return [[SentryCrashReportSink alloc] initWithInAppLogic:self.inAppLogic
                                                crashWrapper:self.crashWrapper];
}

- (void)sendAllReports
{
    [self sendAllReportsWithCompletion:NULL];
}

- (void)sendAllReportsWithCompletion:(SentryCrashReportFilterCompletion)onCompletion
{
    [super
        sendAllReportsWithCompletion:^(NSArray *filteredReports, BOOL completed, NSError *error) {
            if (nil != error) {
                SENTRY_LOG_ERROR(@"%@", error.localizedDescription);
            }
            SENTRY_LOG_DEBUG(@"Sent %lu crash report(s)", (unsigned long)filteredReports.count);
            if (completed && onCompletion) {
                onCompletion(filteredReports, completed, error);
            }
        }];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Integrations/SentryCrash/SentryCrashInstallationReporter.swift
@_implementationOnly import _SentryPrivate
import Foundation

/**
 * Crash installation reporter that handles Sentry-specific reporting details.
 *
 * This class extends SentryCrashInstallation to provide Sentry-specific crash report
 * processing through a SentryCrashReportSink.
 */
final class SentryCrashInstallationReporter: SentryCrashInstallation {

    private let inAppLogic: SentryInAppLogic
    private let crashWrapper: SentryCrashWrapper
    private let dispatchQueue: SentryDispatchQueueWrapper

    init(
        inAppLogic: SentryInAppLogic,
        crashWrapper: SentryCrashWrapper,
        dispatchQueue: SentryDispatchQueueWrapper
    ) {
        self.inAppLogic = inAppLogic
        self.crashWrapper = crashWrapper
        self.dispatchQueue = dispatchQueue
        super.init(requiredProperties: [])
    }

    override func sink() -> (any SentryCrashReportFilter)? {
        return SentryCrashReportSink(
            inAppLogic: inAppLogic,
            crashWrapper: crashWrapper,
            dispatchQueue: dispatchQueue
        )
    }

    override func sendAllReports(completion onCompletion: SentryCrashReportFilterCompletion?) {
        super.sendAllReports { filteredReports, completed, error in
            if let error = error {
                SentrySDKLog.error("Error sending crash reports: \(error.localizedDescription)")
            }
            SentrySDKLog.debug("Sent \(String(describing: filteredReports?.count)) crash report(s)")
            if completed, let onCompletion = onCompletion {
                onCompletion(filteredReports, completed, error)
            }
        }
    }
}
